### PR TITLE
Progress reporting improvements

### LINKED ISSUE #2875: Some filters report progress both from theader and from themselves
### Description
In ITK 5 there's a possibility to report progress from threader. It can be controlled by flag ThreaderUpdateProgressOff() or ThreaderUpdateProgressOn(). 
Right now, 4 filters report their progress both from themselves and from threader, and because of the bug with progress overflow  that I reported earlier (https://github.com/InsightSoftwareConsortium/ITK/issues/2874), the progress overflows: it reaches 100% once, then it goes to 0%, and then again to 100%.
Those filters are:
itk::ResampleImageFilter, itk::LabelMapToLabelImageFilter, itk::LabelShapeKeepNObjectsImageFilter, itk::LabelMapFilter.

Even after the bug with overflow will be fixed, it is still wrong to report progress from both threader and filter, because it will reach 100% before it is actually 100%, so in any case it is needed to add this->ThreaderUpdateProgressOff() in their constructors.

### Steps to Reproduce

Here's the code that reproduces the bug for itk::ResampleImageFilter. The bug is similar for the rest of the filters.
https://gist.github.com/ozavalistyi/9a234eacb36316d922c0a9571ee13db7
### Expected behavior
Progress should be reported up to 100% and then stop.
The output of the code above should not contain "ERROR: Progress overflowed".

### Actual behavior
Progress is reported up to 100% once, then it overflows and is reported from 0% to 100% again (
The output of the above code contains "ERROR: Progress overflowed".
The output is attached to gist, that I provided.

### Reproducibility
100%

### Versions

ITK 5.1.2
But I looked at the latest master, and those filters do not contain this->ThreaderUpdateProgressOff() in their constructors, so it is reproducible with the latest ITK also.
### Environment

Windows 10
Compiler: Visual Studio 2019, version=14.26,host=x64

### PR DESCRIPTION
**Pull request by dzenanz**
Related to #2875.

### CODE REVIEW

**Review by jhlegarreta — COMMENTED** [review]
Have not tried, but wondering whether
> Now the progress fills up twice faster so it reaches 100% when only 50% of work is done.

is the expected behavior, or an expected limitation (e.g. if it cannot be set to 50% when half of the job has been done, and to be 100% only when the job has been completed), whether the wording is accurate or else whether it is a misunderstanding of mine.

File: Modules/Filtering/LabelMap/include/itkLabelMapToLabelImageFilter.hxx [hunk 1]
```diff
@@ -29,6 +29,7 @@ template <typename TInputImage, typename TOutputImage>
 LabelMapToLabelImageFilter<TInputImage, TOutputImage>::LabelMapToLabelImageFilter()
 {
   m_OutputImage = nullptr;
+  this->ThreaderUpdateProgressOff();
```
Comment (blowekamp) [inline Modules/Filtering/LabelMap/include/itkLabelMapToLabelImageFilter.hxx:32]: The base class `LabelMapFilter` is also disabling this. Is this really needed?

File: Modules/Filtering/LabelMap/include/itkLabelShapeKeepNObjectsImageFilter.hxx [hunk 2]
```diff
@@ -29,6 +29,7 @@ LabelShapeKeepNObjectsImageFilter<TInputImage>::LabelShapeKeepNObjectsImageFilte
   m_NumberOfObjects = 1;
   m_ReverseOrdering = false;
   m_Attribute = LabelObjectType::NUMBER_OF_PIXELS;
+  this->ThreaderUpdateProgressOff();
```
Comment (blowekamp) [inline Modules/Filtering/LabelMap/include/itkLabelShapeKeepNObjectsImageFilter.hxx:32]: LabelShaeKeepNObjectsImageFilter is also derived from InPlaceLabelMap/LabelMapFilter

**Comment by dzenanz** [comment]
Before 840bae7 progress would go from 0% to 100% twice, overflowing from 100% to 0% about halfway. After 840bae7, the progress fills up to 100% during first half of the work, and stays there the second half of the work. Now, the progress fills up to 50% during first half of the work, and fills up to 100% during second half of the work.

In all instances, the progress was usually being filled smoothly.

**Review by blowekamp — APPROVED** [review]
Looks good. I just would like to know why it is needed to set is multiple times in a class hierarchy of constructors

File: Modules/Filtering/LabelMap/include/itkLabelMapToLabelImageFilter.hxx (hunk 1 above)
Comment (dzenanz) [reply Modules/Filtering/LabelMap/include/itkLabelMapToLabelImageFilter.hxx:32]: Uh-oh. I didn't look at inheritance hierarchy.

I checked several children of `LabelMapFilter`, they all handle progress reporting themselves. I will remove this duplication.

**Comment by dzenanz** [comment]
@jhlegarreta I updated the commit message.

**Review by jhlegarreta — APPROVED** [review]
Thanks for the patch set and the commit message update, now it makes more sense to me.

After Brad's comment, I'm wondering whether setting/updating the threader update progress report flag should be disabled in child classes/whether that makes sense.

**Comment by dzenanz** [comment]
I have a more comprehensive solution. I am testing it locally.

**Comment by thewtex** [comment]
@dzenanz please start commit messages with capital letters for the ChangeLog.

**Comment by dzenanz** [comment]
There is something wrong with progress updates the second time filter is run:
```text
1570: Trying filter->UpdateLargestPossibleRegion()
1570: -------- Start GradientMagnitudeRecursiveGaussianImageFilter "" Progress Quiet
1570: Filter took 0.0594145 seconds.
1570: -------- End GradientMagnitudeRecursiveGaussianImageFilter ""
1570:
1570: itk::ExceptionObject (000000D8CD2FE3C8)
1570: Location: "void __cdecl itk::SimpleFilterWatcher::EndFilter(void)"
1570: File: C:\Dev\ITK-git\Modules\Core\Common\include\itkSimpleFilterWatcher.h
1570: Line: 283
1570: Description: ITK ERROR: GradientMagnitudeRecursiveGaussianImageFilter(000000D8CD2FED70): Filter does not have progress.
1570:
1570:
1570:   In C:\Dev\ITK-git\Modules\Filtering\ImageGradient\test\itkGradientMagnitudeRecursiveGaussianFilterTest.cxx, line 217
1/1 Test #1570: itkGradientMagnitudeRecursiveGaussianFilterTest ...***Failed    0.47 sec
```

**Comment by dzenanz** [comment]
> something wrong with progress updates the second time filter is run

This is the only problem left. How much of a problem is this? Does someone know how to solve it? I am somewhat tired of this PR. It started as quite a small fix.

**Comment by dzenanz** [comment]
I found out the problem. It will take me some time to clean up the code. I will push the update tomorrow.

**Review by thewtex — APPROVED** [review]

File: Modules/Core/Common/include/itkProgressAccumulator.h [hunk 3]
```diff
@@ -138,6 +138,8 @@ class ITKCommon_EXPORT ProgressAccumulator : public Object
     // The tags for adding/removing observers to mini-pipeline filter
     unsigned long ProgressObserverTag;
     unsigned long StartObserverTag;
+
+    bool AccumulatedProgress{ 0.0 };
```
Comment (blowekamp) [inline Modules/Core/Common/include/itkProgressAccumulator.h:142]: I am getting an error here with conversion from double to bool.

Comment (dzenanz) [reply Modules/Core/Common/include/itkProgressAccumulator.h:142]: I changed the variable name and how it is used in code, but forgot to change the type. And I was wondering why does debugger keep representing it as a float. Silly me! Thank you Brad.

**Comment by blowekamp** [comment]
In testing this PR against SimpleITK I got a large number of FFT related test further inailures, and one Progress report test failing. It will need further investigation.

**Comment by dzenanz** [comment]
Thanks for testing. Feel free to push to this this branch if you find some easy fixes.

**Comment by dzenanz** [comment]
PythonLaplacianImageFilterTest deadlocks on my computer too. At least I can reproduce the failure.

**Comment by dzenanz** [comment]
CI Linux Python:
```text
	209 - PythonGradientMagnitudeRecursiveGaussianImageFilterTest (Timeout)
	210 - itkGradientVectorFlowImageFilterPythonTest (Timeout)
	211 - PythonLaplacianImageFilterTest (Timeout)
	215 - PythonGeodesicActiveContourLeftVentricleTest (Timeout)
```

**Comment by dzenanz** [comment]
I have removed the problematic commit from this PR. I will see if I can iron it out in a follow-up PR. I think this can be merged now.

**Comment by dzenanz** [comment]
I removed all far-reaching code from this PR, and left all the non-problematic code here. It probably no longer fixes #2875, so I changed the description accordingly. I plan the real fix for a follow-up PR.

**Review by jhlegarreta — COMMENTED** [review]
:+1: Thanks @dzenanz. Given that @blowekamp tested a previous version against SimpleITK, may be he would like to take that again, but I'm for merging it as the one-but-last build passed tests, and hope this will as well.

**Comment by blowekamp** [comment]
Having to reset internal filter's progress when re-running seems like a red flag to me. I'd like to test with just re-running a filter in a standalone test to alleviate my concern. I'm not sure I'd have time tomorrow. There are a lot of changes here.

**Review by blowekamp — APPROVED** [review]
I tested a few cases that were of concerned to me, and they all behaved correctly 🥳 

Also SimpleITK tests seemed to be OK too (although there are other FFT issues).

Nice work!

**Comment by dzenanz** [comment]
Are the FFT issues captured in #3048?

**Comment by tbirdso** [comment]
> Also SimpleITK tests seemed to be OK too (although there are other FFT issues).

@blowekamp Could you please add @dzenanz  and myself to FFT discussion if there are other open issues?

**Comment by blowekamp** [comment]
When I narrow in down in SimpleITK, I'll create an issue if needed. Here is a build with the failure FYI: https://open.cdash.org/viewTest.php?onlyfailed&buildid=7657537

**Comment by tbirdso** [comment]
At a glance I'm guessing that the SimpleITK tests will need to add factory registrations for VNL FFT classes as the segfaulting tests all seem to require classes that depend on FFT filter instantiations. Whenever you get around to taking a closer look I'd recommend cross-referencing #3002 to see how this is being done in the ITK TestKernel as well as ITK GTests.

**Pull request merged**